README.md
# ou5
Files used for ou5 in the university course "datatyper och algoritmer (C)" at Umeå university.


changes.txt
v 2.0
Lade till figurer där testkörningar illustreras.
Fixade så att namnet i programbeskrivningen stämmer.
Rätt utskrift vid användande av inkorrekt ''map'' fil.
Inlagd pseudokod.
Ändrade graph_neighbours så att noderna sparas i en separat lista.
Använde string_copy istället för str_cpy.
Fixade minnesläckor.


graph.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "graph.h"
#include "dlist.h"
#include "util.h"

/*
 * Implementation of a generic graph for the "Datastructures and
 * algorithms" courses at the Department of Computing Science, Umea
 * University.
 *
 * Authors: Adam Nilsson (ens17ann@cs.umu.se)
 *	    		Marcus Sundin (tfy18msn@cs.umu.se)
 *
 * Version information:
 *   2021-05-14: v1.0, first public version.
 *   2021-05-27: v2.0, small changes.
 *   2021-06-01: v3.0, memory leaks fixed.
 */

// ===========INTERNAL DATA TYPES============

struct graph {
	dlist *nodes;
	int nmrOfNodes;
};

struct node {
  dlist *edges;
  char *name;
  bool seen;
};

// ===========INTERNAL FUNCTION IMPLEMENTATIONS============

/**
 * nodes_are_equal() - Check whether two nodes are equal.
 * @n1: Pointer to node 1.
 * @n2: Pointer to node 2.
 *
 * Returns: true if the nodes are considered equal, otherwise false.
 *
 */
bool nodes_are_equal(const node *n1,const node *n2)
{
  return n1->name == n2->name;
}

// =================== GRAPH STRUCTURE INTERFACE ======================

/**
 * graph_empty() - Create an empty graph.
 * @max_nodes: The maximum number of nodes the graph can hold.
 *
 * Returns: A pointer to the new graph.
 */
graph *graph_empty(int max_nodes)
{
  // Allocate the graph header.
  struct graph *g = calloc(1, sizeof(graph));
  // Create the lists to hold the nodes and edges.
  g->nodes = dlist_empty(NULL);
	g->nmrOfNodes = 0;

  return g;
}

/**
 * graph_is_empty() - Check if a graph is empty, i.e. has no nodes.
 * @g: Graph to check.
 *
 * Returns: True if graph is empty, otherwise false.
 */
bool graph_is_empty(const graph *g)
{
	return g->nmrOfNodes == 0;
}

/**
 * graph_insert_node() - Inserts a node with the given name into the graph.
 * @g: Graph to manipulate.
 * @s: Node name.
 *
 * Creates a new node with a copy of the given name and puts it into
 * the graph.
 *
 * Returns: The modified graph.
 */
 graph *graph_insert_node(graph *g, const char *s)
 {
 	struct node *n = calloc(1,sizeof(*n));
 	n->edges = dlist_empty(NULL);
 	n->name = (char*) s;
 	n->seen = false;
 	dlist_insert(g->nodes, n, dlist_first(g->nodes));
 	g->nmrOfNodes++;

 	return g;
 }

/**
 * graph_find_node() - Find a node stored in the graph.
 * @g: Graph to manipulate.
 * @s: Node identifier, e.g. a char *.
 *
 * Returns: A pointer to the found node, or NULL.
 */
node *graph_find_node(const graph *g, const char *s)
{
  dlist_pos pos = dlist_first(g->nodes);
  while (!dlist_is_end(g->nodes, pos)) {
    struct node *n = dlist_inspect(g->nodes, pos);
	 	char *name = n->name;
    if (!strcmp(name, s)) {
      return n;
    }
    pos = dlist_next(g->nodes, pos);
  }
  return NULL;
}

/**
 * graph_node_is_seen() - Return the seen status for a node.
 * @g: Graph storing the node.
 * @n: Node in the graph to return seen status for.
 *
 * Returns: The seen status for the node.
 */
bool graph_node_is_seen(const graph *g, const node *n)
{
	return n->seen;
}

/**
 * graph_node_set_seen() - Set the seen status for a node.
 * @g: Graph storing the node.
 * @n: Node in the graph to set seen status for.
 * @s: Status to set.
 *
 * Returns: The modified graph.
 */
graph *graph_node_set_seen(graph *g, node *n, bool seen)
{
  n->seen = seen;
  return g;
}

/**
 * graph_reset_seen() - Reset the seen status on all nodes in the graph.
 * @g: Graph to modify.
 *
 * Returns: The modified graph.
 */
graph *graph_reset_seen(graph *g)
{
	dlist_pos pos = dlist_first(g->nodes);

	while (!dlist_is_end(g->nodes, pos))
	{
		struct node *n = dlist_inspect(g->nodes,pos);
		n->seen = false;

		pos = dlist_next(g->nodes, pos);
	}
	return g;
}

/**
 * graph_insert_edge() - Insert an edge into the graph.
 * @g: Graph to manipulate.
 * @n1: Source node (pointer) for the edge.
 * @n2: Destination node (pointer) for the edge.
 *
 * NOTE: Undefined unless both nodes are already in the graph.
 *
 * Returns: The modified graph.
 */
graph *graph_insert_edge(graph *g, node *n1, node *n2)
{
  dlist_insert(n1->edges, n2->name, dlist_first(n1->edges));
  return g;
}

/**
 * graph_neighbours() - Return a list of neighbour nodes.
 * @g: Graph to inspect.
 * @n: Node to get neighbours for.
 *
 * Returns: A pointer to a list of nodes. Note: The list must be
 * dlist_kill()-ed after use.
 */
 dlist *graph_neighbours(const graph *g,const node *n)
 {
	 struct dlist *neighbours = dlist_empty(NULL);
	 //neighbours= n->edges;

	dlist_pos pos = dlist_first(n->edges);
	dlist_pos pos2 = dlist_first(neighbours);

	while (!dlist_is_end(n->edges, pos)){

	 	dlist_insert(neighbours, dlist_inspect(n->edges,pos), pos2);
	 	pos = dlist_next(n->edges, pos);
	 	pos2 = dlist_next(neighbours, pos2);
	 }
 	return neighbours;
}

/**
 * graph_kill() - Destroy a given graph.
 * @g: Graph to destroy.
 *
 * Return all dynamic memory used by the graph.
 *
 * Returns: Nothing.
 */

void graph_kill(graph *g)
{
	// Iterate over the list. Destroy all elements.
	dlist_pos pos1 = dlist_first(g->nodes);

	while (!dlist_is_end(g->nodes, pos1)) {
		struct node *n = dlist_inspect(g->nodes, pos1);

		//Deallocate the list where the nodes are saved.
		dlist_kill(n->edges);
		free(n->name);
		// Move on to next element.
		pos1 = dlist_next(g->nodes, pos1);
    // Deallocate the table entry structure.
    free(n);
	}

	// Kill what's left of the list...
	dlist_kill(g->nodes);
	// ...and the table.
	free(g);
}


graph.h
#ifndef __GRAPH_H
#define __GRAPH_H

#include <stdbool.h>
#include "util.h"
#include "dlist.h"

/*
 * Declaration of a generic graph for the "Datastructures and
 * algorithms" courses at the Department of Computing Science, Umea
 * University. The graph stores nodes and edges of a directed or
 * undirected graph. After use, the function graph_kill() must
 * be called to de-allocate the dynamic memory used by the graph
 * itself. The de-allocation of any dynamic memory allocated for the
 * node names is the responsibility of the user of the graph.
 *
 * Author: Niclas Borlin (niclas.borlin@cs.umu.se)
 *
 * Version information:
 *   v1.0  2019-02-21: First public version.

 *   v1.01 2019-02-26: Modified include directives for header files
 *                     from the codebase to use "" to handle case when
 *                     student stores all header files in the current
 *                     directory. See
 *                     https://stackoverflow.com/questions/21593/what-is-the-difference-between-include-filename-and-include-filename.
 *   v1.02 2019-03-01: Doc update in graph_node_is_seen().
 *   v1.1  2019-03-06: Changed several const node * to node *.
 *                     Fixed doc bug to state that any dynamic memory allocated
 *                     to the node NAMES is the resposibility of the user.
 */

// ====================== PUBLIC DATA TYPES ==========================

// Anonymous declarations of node and graph.
typedef struct node node;
typedef struct graph graph;

// =================== NODE COMPARISON FUNCTION ======================

/**
 * nodes_are_equal() - Check whether two nodes are equal.
 * @n1: Pointer to node 1.
 * @n2: Pointer to node 2.
 *
 * Returns: true if the nodes are considered equal, otherwise false.
 *
 */
bool nodes_are_equal(const node *n1,const node *n2);

// =================== GRAPH STRUCTURE INTERFACE ======================

/**
 * graph_empty() - Create an empty graph.
 * @max_nodes: The maximum number of nodes the graph can hold.
 *
 * Returns: A pointer to the new graph.
 */
graph *graph_empty(int max_nodes);

/**
 * graph_is_empty() - Check if a graph is empty, i.e. has no nodes.
 * @g: Graph to check.
 *
 * Returns: True if graph is empty, otherwise false.
 */
bool graph_is_empty(const graph *g);

/**
 * graph_has_edges() - Check if a graph has any edges.
 * @g: Graph to check.
 *
 * Returns: True if graph has any edges, otherwise false.
 */
bool graph_has_edges(const graph *g);

/**
 * graph_insert_node() - Inserts a node with the given name into the graph.
 * @g: Graph to manipulate.
 * @s: Node name.
 *
 * Creates a new node with a copy of the given name and puts it into
 * the graph.
 *
 * Returns: The modified graph.
 */
graph *graph_insert_node(graph *g, const char *s);

/**
 * graph_find_node() - Find a node stored in the graph.
 * @g: Graph to manipulate.
 * @s: Node identifier, e.g. a char *.
 *
 * Returns: A pointer to the found node, or NULL.
 */
node *graph_find_node(const graph *g, const char *s);

/**
 * graph_node_is_seen() - Return the seen status for a node.
 * @g: Graph storing the node.
 * @n: Node in the graph to return seen status for.
 *
 * Returns: The seen status for the node.
 */
bool graph_node_is_seen(const graph *g, const node *n);

/**
 * graph_node_set_seen() - Set the seen status for a node.
 * @g: Graph storing the node.
 * @n: Node in the graph to set seen status for.
 * @s: Status to set.
 *
 * Returns: The modified graph.
 */
graph *graph_node_set_seen(graph *g, node *n, bool seen);

/**
 * graph_reset_seen() - Reset the seen status on all nodes in the graph.
 * @g: Graph to modify.
 *
 * Returns: The modified graph.
 */
graph *graph_reset_seen(graph *g);

/**
 * graph_insert_edge() - Insert an edge into the graph.
 * @g: Graph to manipulate.
 * @n1: Source node (pointer) for the edge.
 * @n2: Destination node (pointer) for the edge.
 *
 * NOTE: Undefined unless both nodes are already in the graph.
 *
 * Returns: The modified graph.
 */
graph *graph_insert_edge(graph *g, node *n1, node *n2);

/**
 * graph_neighbours() - Return a list of neighbour nodes.
 * @g: Graph to inspect.
 * @n: Node to get neighbours for.
 *
 * Returns: A pointer to a list of nodes. Note: The list must be
 * dlist_kill()-ed after use.
 */
dlist *graph_neighbours(const graph *g,const node *n);

/**
 * graph_kill() - Destroy a given graph.
 * @g: Graph to destroy.
 *
 * Return all dynamic memory used by the graph.
 *
 * Returns: Nothing.
 */
void graph_kill(graph *g);

/**
 * graph_print() - Iterate over the graph elements and print their values.
 * @g: Graph to inspect.
 *
 * Iterates over the graph and prints its contents.
 *
 * Returns: Nothing.
 */
void graph_print(const graph *g);

#endif


graph_local.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "graph.h"
#include "dlist.h"
#include "util.h"

/*
 * Implementation of a generic graph for the "Datastructures and
 * algorithms" courses at the Department of Computing Science, Umea
 * University.
 *
 * Authors: Adam Nilsson (ens17ann@cs.umu.se)
 *	    		Marcus Sundin (tfy18msn@cs.umu.se)
 *
 * Version information:
 *   2021-05-14: v1.0, first public version.
 *   2021-05-27: v2.0, small changes.
 */

// ===========INTERNAL DATA TYPES============

struct graph {
	dlist *nodes;
	int nmrOfNodes;
};

struct node {
  dlist *edges;
  char *name;
  bool seen;
};

// ===========INTERNAL FUNCTION IMPLEMENTATIONS============

/**
 * nodes_are_equal() - Check whether two nodes are equal.
 * @n1: Pointer to node 1.
 * @n2: Pointer to node 2.
 *
 * Returns: true if the nodes are considered equal, otherwise false.
 *
 */
bool nodes_are_equal(const node *n1,const node *n2)
{
  return n1->name == n2->name;
}

// =================== GRAPH STRUCTURE INTERFACE ======================

/**
 * graph_empty() - Create an empty graph.
 * @max_nodes: The maximum number of nodes the graph can hold.
 *
 * Returns: A pointer to the new graph.
 */
graph *graph_empty(int max_nodes)
{
  // Allocate the graph header.
  struct graph *g = calloc(1, sizeof(graph));
  // Create the lists to hold the nodes and edges.
  g->nodes = dlist_empty(NULL);
	g->nmrOfNodes = 0;

  return g;
}

/**
 * graph_is_empty() - Check if a graph is empty, i.e. has no nodes.
 * @g: Graph to check.
 *
 * Returns: True if graph is empty, otherwise false.
 */
bool graph_is_empty(const graph *g)
{
	return g->nmrOfNodes == 0;
}

/**
 * graph_insert_node() - Inserts a node with the given name into the graph.
 * @g: Graph to manipulate.
 * @s: Node name.
 *
 * Creates a new node with a copy of the given name and puts it into
 * the graph.
 *
 * Returns: The modified graph.
 */
 graph *graph_insert_node(graph *g, const char *s)
 {
 	struct node *n = calloc(1,sizeof(*n));
 	n->edges = dlist_empty(NULL);
 	n->name = (char*) s;
 	n->seen = false;
 	dlist_insert(g->nodes, n, dlist_first(g->nodes));
 	g->nmrOfNodes++;

 	return g;
 }

/**
 * graph_find_node() - Find a node stored in the graph.
 * @g: Graph to manipulate.
 * @s: Node identifier, e.g. a char *.
 *
 * Returns: A pointer to the found node, or NULL.
 */
node *graph_find_node(const graph *g, const char *s)
{
  dlist_pos pos = dlist_first(g->nodes);
  while (!dlist_is_end(g->nodes, pos)) {
    struct node *n = dlist_inspect(g->nodes, pos);
	 	char *name = n->name;
    if (!strcmp(name, s)) {
      return n;
    }
    pos = dlist_next(g->nodes, pos);
  }
  return NULL;
}

/**
 * graph_node_is_seen() - Return the seen status for a node.
 * @g: Graph storing the node.
 * @n: Node in the graph to return seen status for.
 *
 * Returns: The seen status for the node.
 */
bool graph_node_is_seen(const graph *g, const node *n)
{
	return n->seen;
}

/**
 * graph_node_set_seen() - Set the seen status for a node.
 * @g: Graph storing the node.
 * @n: Node in the graph to set seen status for.
 * @s: Status to set.
 *
 * Returns: The modified graph.
 */
graph *graph_node_set_seen(graph *g, node *n, bool seen)
{
  n->seen = seen;
  return g;
}

/**
 * graph_reset_seen() - Reset the seen status on all nodes in the graph.
 * @g: Graph to modify.
 *
 * Returns: The modified graph.
 */
graph *graph_reset_seen(graph *g)
{
	dlist_pos pos = dlist_first(g->nodes);

	while (!dlist_is_end(g->nodes, pos))
	{
		struct node *n = dlist_inspect(g->nodes,pos);
		n->seen = false;

		pos = dlist_next(g->nodes, pos);
	}
	return g;
}

/**
 * graph_insert_edge() - Insert an edge into the graph.
 * @g: Graph to manipulate.
 * @n1: Source node (pointer) for the edge.
 * @n2: Destination node (pointer) for the edge.
 *
 * NOTE: Undefined unless both nodes are already in the graph.
 *
 * Returns: The modified graph.
 */
graph *graph_insert_edge(graph *g, node *n1, node *n2)
{
  dlist_insert(n1->edges, n2->name, dlist_first(n1->edges));
  return g;
}

/**
 * graph_neighbours() - Return a list of neighbour nodes.
 * @g: Graph to inspect.
 * @n: Node to get neighbours for.
 *
 * Returns: A pointer to a list of nodes. Note: The list must be
 * dlist_kill()-ed after use.
 */
 dlist *graph_neighbours(const graph *g,const node *n)
 {
	 struct dlist *neighbours = dlist_empty(NULL);
	 //neighbours= n->edges;

	dlist_pos pos = dlist_first(n->edges);
	dlist_pos pos2 = dlist_first(neighbours);

	while (!dlist_is_end(n->edges, pos)){

	 	dlist_insert(neighbours, dlist_inspect(n->edges,pos), pos2);
	 	pos = dlist_next(n->edges, pos);
	 	pos2 = dlist_next(neighbours, pos2);
	 }
 	return neighbours;
}

/**
 * graph_kill() - Destroy a given graph.
 * @g: Graph to destroy.
 *
 * Return all dynamic memory used by the graph.
 *
 * Returns: Nothing.
 */

void graph_kill(graph *g)
{
	// Iterate over the list. Destroy all elements.
	dlist_pos pos1 = dlist_first(g->nodes);

	while (!dlist_is_end(g->nodes, pos1)) {
		struct node *n = dlist_inspect(g->nodes, pos1);

		//Deallocate the list where the nodes are saved.
		dlist_kill(n->edges);

		// Move on to next element.
		pos1 = dlist_next(g->nodes, pos1);
    // Deallocate the table entry structure.
    free(n);
	}

	// Kill what's left of the list...
	dlist_kill(g->nodes);
	// ...and the table.
	free(g);
}


is_connected_new.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#include <string.h>
#include "queue.h"
#include "graph.h"
#include "dlist.h"

#define MAXCHAR 1000

/*
 * Implementation of a breadth-first algorithm to determine if a path is connected.
 * The program first scans a map file to create a graph.
 * The graph is then traversed with a breadth-first algorihm.
 *
 * Authors: Adam Nilsson (ens17ann@cs.umu.se)
 *	        Marcus Sundin (tfy18msn@cs.umu.se)
 *
 * Version information:
 *   2021-05-16: v1.0, first public version.
 */

 /**
 * copy_string() - Create a dynamic copy of a string.
 * @s: String to be copied.
 *
 * Allocates memory for a dynamic copy of s and copies the contents of
 * s into the copy.
 *
 * Returns: Pointer to the copy of s.
 */
char *copy_string(const char *s)
{
  int len=strlen(s);

  /* Allocate memory for new string, with an extra char for \0 */
  char *dest = malloc(sizeof(char)*(len+1));

  /* Malloc failed, return NULL */
  if (dest == NULL) {
    return NULL;
  }

  /* Copy content to new memory */
  strncpy(dest, s, len);

  /* Strings should always be null terminated */
  dest[len] = '\0';
  return dest;
}

 bool find_path(graph *g, node *src, node *dest)
 {
   //Creating an empty queue
   queue *q = queue_empty(NULL);
   bool seen = true;
   {
     if (nodes_are_equal(dest,src))
     {
       return true;
     }
   }

   graph_node_set_seen(g, src, seen); // Mark the "src" node as seen.
   queue_enqueue(q,src); // Put "src" node in queue.
   //Check if the nodes are the same, returns that the nodes are connected if true.

   while (!queue_is_empty(q))
   {

     struct node *p = queue_front(q); // Pick first node from queue
     queue_dequeue(q); //Remove the first element from the queue
     struct dlist *neighbours=graph_neighbours(g,p);//saving the neighbours in a directed list.
     dlist_pos pos = dlist_first(neighbours); //checking the position of first element.
     //Go through neighbours and add to queue if the node is not set to seen.

     while(!dlist_is_end(neighbours,pos))
     {
       char* name = dlist_inspect(neighbours, pos);
       struct node *m = graph_find_node(g, name);
       //Check if the node "m" is set as seen.
       if (!graph_node_is_seen(g, m))
       {
         //Check if the node "m" is the same as "dest",
         //returns that the nodes are connected if true.
         if (nodes_are_equal(m,dest))
         {
           while (!queue_is_empty(q))
           {
             queue_dequeue(q);
           }
           queue_kill(q);
           return true;
         }
         //If nodes  are not equal continue with the traversing by setting the node as seen.
         //The node is also added into the queue.
         else
         {
           g = graph_node_set_seen(g, m, seen); //Set the node "m" as seen.
           q = queue_enqueue(q,m); //Add the node "m" to the queue.
         }
       }
       pos = dlist_next(neighbours,pos); //Save the position of the next element.
     }
   }
   queue_kill(q);

   return false; //if "dest" not found return false
 }

int main(int argc, char** argv) {

  FILE *fp;
  char str[MAXCHAR];
  char* filename = argv[argc-1];

  fp = fopen(filename, "r");  //Open mapfile

  int i = 0;
  int number_of_edges = -1;
  bool error = false;         //Checks if everything works correctly
  graph *g = graph_empty(100); //empty graph

  //string vector containing nodes connected by edges
  while (fgets(str, MAXCHAR, fp) != NULL) {
    char* origin;
    char* destination;

    if (str[0] != '#' && str[0] != ' ' && number_of_edges == -1) {
      number_of_edges = atoi(str);
    }

    //ignors rows starting with # or blanck space
    else if (str[0] != '#' && str[0] != ' ')
    {
      char sub[41]; //name holder variable
      int j = 0;
      //reads the name until it hits a blanck space
      while(str[j] != ' ' && !error)
      {
        sub[j] = str[j];
        if (str[j] == '\n') //error if it hits a new row already
          error = true;
        j++;
      }
      if (error) {
        break;
      }
      sub[j] = '\0';
      origin = copy_string(sub);

      j++;
      int k = 0;
      //reads the name until it ends
      while(str[j] != ' ' && str[j] != '#' && str[j] != '\n') {
        sub[k] = str[j];
        j++;
        k++;
      }
      sub[k] = '\0';
      destination = copy_string(sub);
      i++;

      //inserts the origin nodes
      if (graph_find_node(g, origin) == NULL) {
        graph_insert_node(g, origin);
      }
      //inserts the destination nodes
      if (graph_find_node(g, destination) == NULL) {
        graph_insert_node(g, destination);
      }
      //inserts the edges
      struct node *n1 = graph_find_node(g, origin);
      struct node *n2 = graph_find_node(g, destination);
      graph_insert_edge(g, n1, n2);
    }
  }

  //checks if number_of_edges has been given a value
  if (number_of_edges < 1)
    error = true;

  //checks that the indicated number of edges is correct
  if (i != number_of_edges)
    error = true;

  //error message
  if (error) {
    printf("Invalid map\n");
    exit(EXIT_FAILURE);
  }

  fclose(fp);

  bool loop=true;
  while (loop)
  {
    // char* src_name;
    // char* dest_name;
    char names[2][41];
    error = false;  //reset error
    //asks for node names
    printf("Enter origin and destination (quit to exit): ");
    for (int i = 0; i < 2; i++) {
      scanf("%s", names[i]);
      //checks if user has entered "quit"
      if (!strcmp(names[0], "quit")) {
        loop = false;
        break;
      }
    }

    if (loop)
    {
      // src_name = copy_string(names[0]);
      // dest_name = copy_string(names[1]);
      //checks if nodes exist
      if (graph_find_node(g, names[0]) == NULL ||
          graph_find_node(g, names[1]) == NULL)
          error = true;

      if (!error) {
        struct node *src = graph_find_node(g,  names[0]);
        struct node *dest = graph_find_node(g,  names[1]);

        bool connected = find_path(g, src, dest); //try to find a path
        //message
        if (connected)
        {
          printf("There is a path from %s to %s.\n",  names[0],  names[1]);
        }
        else
        {
          printf("There is no path from %s to %s.\n",  names[0],  names[1]);
        }
        //reset seen status between runs:
        g = graph_reset_seen(g);
      }
      else
        printf("Invalid input\n");
    }
  }
  graph_kill(g);  //kill the graph
  printf("Normal exit.\n");
  return 0;
}


is_connected_file_reader.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
//#include "queue.h"
#include "graph.h"
#include "dlist.h"

#define MAXCHAR 1000

int main(int argc, char** argv) {

  FILE *fp;
  char str[MAXCHAR];
  char* filename = argv[argc-1];

  fp = fopen(filename, "r");

  int i = 0;
  int number_of_edges = -1;

  while (fgets(str, MAXCHAR, fp) != NULL) {
    if (str[0] != '#' && str[0] != ' ') {
      number_of_edges = atoi(str);
      break;
    }
  }
  char origins[number_of_edges][41];
  char destinations[number_of_edges][41];

  while (fgets(str, MAXCHAR, fp) != NULL) {
    if (str[0] != '#' && str[0] != ' ') {
      char sub[41];
      int j = 0;
      while(str[j] != ' ') {
        sub[j] = str[j];
        j++;
      }
      strcpy(origins[i], sub);
      memset(sub, 0, strlen(sub));

      j++;
      int k = 0;
      while(str[j] != ' ' && str[j] != '#' && str[j] != '\n') {
        sub[k] = str[j];
        j++;
        k++;
      }
      strcpy(destinations[i], sub);
      memset(sub, 0, strlen(sub));
      i++;
    }
  }

  fclose(fp);

  graph *g = graph_empty(10);

  for (size_t i = 0; i < number_of_edges; i++) {
    if (graph_find_node(g, origins[i]) == NULL) {
      graph_insert_node(g, origins[i]);
    }
  }

  for (size_t i = 0; i < number_of_edges; i++) {
    if (graph_find_node(g, destinations[i]) == NULL) {
      graph_insert_node(g, destinations[i]);
    }
  }

  for (size_t i = 0; i < number_of_edges; i++) {
    struct node *n1 = graph_find_node(g, origins[i]);
    struct node *n2 = graph_find_node(g, destinations[i]);
    graph_insert_edge(g, n1, n2);
  }

  return 0;
}


is_connected.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#include <string.h>
#include "queue.h"
#include "graph.h"
#include "dlist.h"

#define MAXCHAR 1000

/*
 * Implementation of a breadth-first algorithm to determine if a path is connected.
 * The program first scans a map file to create a graph.
 * The graph is then traversed with a breadth-first algorihm.
 *
 * Authors: Adam Nilsson (ens17ann@cs.umu.se)
 *	        Marcus Sundin (tfy18msn@cs.umu.se)
 *
 * Version information:
 *   2021-05-16: v1.0, first public version.
 */

 /**
 * copy_string() - Create a dynamic copy of a string.
 * @s: String to be copied.
 *
 * Allocates memory for a dynamic copy of s and copies the contents of
 * s into the copy.
 *
 * Returns: Pointer to the copy of s.
 */
char *copy_string(const char *s)
{
  int len=strlen(s);

  /* Allocate memory for new string, with an extra char for \0 */
  char *dest = malloc(sizeof(char)*(len+1));

  /* Malloc failed, return NULL */
  if (dest == NULL) {
    return NULL;
  }

  /* Copy content to new memory */
  strncpy(dest, s, len);

  /* Strings should always be null terminated */
  dest[len] = '\0';
  return dest;
}

 bool find_path(graph *g, node *src, node *dest)
 {
   //Creating an empty queue
   queue *q = queue_empty(NULL);
   bool seen = true;

   if (nodes_are_equal(dest,src))
   {
     queue_kill(q);
     return true;
   }

   graph_node_set_seen(g, src, seen); // Mark the "src" node as seen.
   queue_enqueue(q,src); // Put "src" node in queue.
   //Check if the nodes are the same, returns that the nodes are connected if true.

   while (!queue_is_empty(q))
   {

     struct node *p = queue_front(q); // Pick first node from queue
     queue_dequeue(q); //Remove the first element from the queue
     struct dlist *neighbours=graph_neighbours(g,p);//saving the neighbours in a directed list.
     dlist_pos pos = dlist_first(neighbours); //checking the position of first element.
     //Go through neighbours and add to queue if the node is not set to seen.

     while(!dlist_is_end(neighbours,pos))
     {
       char* name = dlist_inspect(neighbours, pos);
       struct node *m = graph_find_node(g, name);
       //Check if the node "m" is set as seen.
       if (!graph_node_is_seen(g, m))
       {
         //Check if the node "m" is the same as "dest",
         //returns that the nodes are connected if true.
         if (nodes_are_equal(m,dest))
         {
           while (!queue_is_empty(q))
           {
             queue_dequeue(q);
           }
           queue_kill(q);
           dlist_kill(neighbours);
           return true;
         }
         //If nodes  are not equal continue with the traversing by setting the node as seen.
         //The node is also added into the queue.
         else
         {
           g = graph_node_set_seen(g, m, seen); //Set the node "m" as seen.
           q = queue_enqueue(q,m); //Add the node "m" to the queue.
         }
       }
       pos = dlist_next(neighbours,pos); //Save the position of the next element.
     }
   dlist_kill(neighbours);
   }
   queue_kill(q);


   return false; //if "dest" not found return false
 }

int main(int argc, char** argv) {

  FILE *fp;
  char str[MAXCHAR];
  char* filename = argv[argc-1];

  fp = fopen(filename, "r");  //Open mapfile

  int i = 0;
  int number_of_edges = -1;
  bool error = false;         //Checks if everything works correctly
  graph *g = graph_empty(100); //empty graph
  if (argc == 1) {
    printf("No map!\n");
    exit(EXIT_FAILURE);
  }

  //string vector containing nodes connected by edges
  while (fgets(str, MAXCHAR, fp) != NULL) {

    if (str[0] != '#' && str[0] != ' ' && str[0] != '\n' &&
        number_of_edges == -1) {
      number_of_edges = atoi(str);
    }

    //ignors rows starting with # or blanck space
    else if (str[0] != '#' && str[0] != ' ')
    {
      char sub[41]; //name holder variable
      int j = 0;
      //reads the name until it hits a blanck space
      while(str[j] != ' ' && !error)
      {
        sub[j] = str[j];
        if (str[j] == '\n') //error if it hits a new row already
          error = true;
        j++;
      }
      if (error) {
        break;
      }
      sub[j] = '\0';
      char* origin = copy_string(sub);

      j++;
      int k = 0;
      //reads the name until it ends
      while(str[j] != ' ' && str[j] != '#' && str[j] != '\n') {
        sub[k] = str[j];
        j++;
        k++;
      }
      sub[k] = '\0';
      char* destination = copy_string(sub);
      i++;

      bool free_origin = false;
      bool free_destination = false;
      //inserts the origin nodes
      if (graph_find_node(g, origin) == NULL) {
        graph_insert_node(g, origin);
      }
      else
        free_origin = true;
      //inserts the destination nodes
      if (graph_find_node(g, destination) == NULL) {
        graph_insert_node(g, destination);
      }
      else
        free_destination = true;
      //inserts the edges
      struct node *n1 = graph_find_node(g, origin);
      struct node *n2 = graph_find_node(g, destination);
      graph_insert_edge(g, n1, n2);

      if (free_origin)
        free(origin);
      if (free_destination)
        free(destination);
    }
  }
  //checks if number_of_edges has been given a value
  if (number_of_edges < 1)
    error = true;

  //checks that the indicated number of edges is correct
  if (i != number_of_edges)
    error = true;

  //error message
  if (error) {
    printf("Invalid map\n");
    exit(EXIT_FAILURE);
  }

  fclose(fp);

  bool loop=true;
  while (loop)
  {
    // char* src_name;
    // char* dest_name;
    char names[2][41];
    error = false;  //reset error
    //asks for node names
    printf("Enter origin and destination (quit to exit): ");
    for (int i = 0; i < 2; i++) {
      scanf("%s", names[i]);
      //checks if user has entered "quit"
      if (!strcmp(names[0], "quit")) {
        loop = false;
        break;
      }
    }

    if (loop)
    {
      // src_name = copy_string(names[0]);
      // dest_name = copy_string(names[1]);
      //checks if nodes exist
      if (graph_find_node(g, names[0]) == NULL ||
          graph_find_node(g, names[1]) == NULL)
          error = true;

      if (!error) {
        struct node *src = graph_find_node(g,  names[0]);
        struct node *dest = graph_find_node(g,  names[1]);

        bool connected = find_path(g, src, dest); //try to find a path
        //message
        if (connected)
        {
          printf("There is a path from %s to %s.\n",  names[0],  names[1]);
        }
        else
        {
          printf("There is no path from %s to %s.\n",  names[0],  names[1]);
        }
        //reset seen status between runs:
        g = graph_reset_seen(g);
      }
      else
      printf("Invalid input\n");
    }
  }
  graph_kill(g);  //kill the graph
  printf("Normal exit.\n");
  return 0;
}
